man/man3/initdrain.3
.TH initdrain 3  "LPJmL programmers manual"
.SH NAME
initdrain \- initialize river routing network
.SH SYNOPSIS
.nf
\fB#include "lpj.h"

Bool initdrain(Cell \fIgrid\fB[],Config *\fIconfig\fB);\fP

.fi
.SH DESCRIPTION
Function initializes river routing network.  
The arguments of \fBinitdrain\fP are:
.TP
.I grid
Cell grid array initialized by \fBnewgrid()\fP.
.TP
.I config
Pointer to LPJmL configuration data. Must be read by \fBreadconfig()\fP.
.SH RETURN VALUE
Upon successful completion FALSE is returned and TRUE in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
newgrid(3), readconfig(3), iterate(3) 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

## [5.7.xx] - 2023-08-xx

### Removed

- removed large chunks of code in `allocation_daily_crop.c` that was inactive

### Fixed

- fixed output `PFT_VEGC` to not include `bm_inc.nitrogen` for crops in `daily_agriculture.c`

## [5.6.25] - 2023-07-29

### Added

- New outputs added for N fluxes from timber extraction and for N pools in wood product pools
- New outputs added for `estab_storage` C and N
- Added missing N fluxes, `estab_storage`, and reservoirs to `globalflux` output
- Added new fluxes to man pages

### Fixed 

- Corrected/added units in comments in output.h
- Corrected units in `flux_sum` man page
- Corrected `N2O_DENIT` output, which now no longer also includes `N2O_NIT` values
- Corrected writing of mineral N pools to outputs, so that also mineral N in frozen soil layers is included
- Corrected writing of `VEGN` for crops, so that `bm_inc.nitrogen` is no longer included to avoid double accounting

### Removed

- Removed unecessary initialization of output variables to zero in `update_annual.c`, which is done in `initoutputdata.c`


## [5.6.24] - 2023-07-29

### Fixed

-  The number of bands is now correctly checked for the `read_residue_data` input.


## [5.6.23] - 2023-07-28

### Changed

- Instead of the soil code the index of the grid cell is written into the NetCDF file for `"grid"` output.


## [5.6.22] - 2023-07-27

### Changed

- Configuration files `lpjml_netcdf.js`, `lpjml_non.js`, `param_non.js`, and `lpjparam_non.js` updated to latest lpjml version.

### Removed

- Predefined daily outputs (`"d_lai"`-`"d_pet"`) and corresponding functions removed.

- Version numbers removed from man pages.

### Fixed

- Variable `issocket` always initialized to `FALSE` in `readfilename.c`.


## [5.6.21] - 2023-07-21

### Changed

- Flag `"nitrogen_coupled"` disabled for lpjml offline runs. Water stress is always coupled with nitrogen stress. Only if lpjml is compiled with `-DCOUPLING_WITH_FMS` flag is enabled.

- Flag `"new_trf"` renamed to `"transp_suction_fcn"`.


## [5.6.20] - 2023-07-13

### Added

- New flags `"fix_co2"` and `"fix_co2_year"` added.

### Changed

- Flag `"istimber"` renamed to `"luc_timber"`.

- Flag `"new_phenology"` renamed to `"gsi_phenology"`.

- Flag `"shuffle_climate"` renamed to `"shuffle_spinup_climate"`.

- The settings for fixed input data have been updated. For climate and N deposition input an interval can be specified from which data is taken after the specified year. In can specified whether random shuffling or cycling is used:

```java
"fix_climate" : false,                /* fix climate after specified year */
"fix_climate_year" : 1901,            /* year after climate is fixed */
"fix_climate_interval" : [1901,1930],
"fix_climate_shuffle" : true,          /* randomly shuffle climate in the interval */
"fix_deposition_with_climate" : false, /* fix N deposition same as climate */
"fix_deposition" : false,              /* fix N deposition after specified year */
"fix_deposition_year" : 1901,          /* year after deposition is fixed */
"fix_deposition_interval" : [1901,1930],
"fix_deposition_shuffle" : true,       /* randomly shuffle depositions in the interval */
"fix_landuse" : false,                 /* fix land use after specfied year */
"fix_landuse_year" : 1901,             /* year after land use is fixed */
"fix_co2" : false,                     /* fix atmospheric CO2  after specfied year */
"fix_co2_year" : 1901,                 /* year after CO2 is fixed */
```

### Removed

- Obsolete flag `"const_climate"` removed.


## [5.6.19] - 2023-07-06

### Added

- Help option `-h` added to `bin2cdf` and `clm2cdf` utility.

- Target `lpjcheck` added to `Makefile` to create only the `lpjcheck` utility.

### Changed

- The version of all man pages has been set to the LPJmL version number, date of last modification removed.

- The different versions of the utilities have been replaced by a reference to the LPJmL version.

### Removed

- All html files have been removed from the repository.

- Obsolete link `lpj` to `lpjml` removed.


## [5.6.18] - 2023-07-04

### Fixed

- Soil array `wpwp` is now correctly initialized with `soilpar->wpwp` for the `"prescribed_soilpar"` setting.


## [5.6.17] - 2023-06-30

### Fixed

- Replaced logical `or` by logical `and` in `survive.c`. This solves issue [#309](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/309).


## [5.6.16] - 2023-06-30

### Fixed

- Removed excess `pft->nind` in all harvest routines in `harvest_stand.c`.


## [5.6.15] - 2023-06-30

### Added

- Check for matching size of discharge queues read from restart file added to `initriver()` function in `initdrain.c`. If the size differs from the value calculated from the river lengths the queues will be resized and initialized to zero. Error message is additionally printed.


## [5.6.14] - 2023-06-08

### Removed

- Support for the deprecated `*.conf` configuration file format has been removed. Only JSON files are allowed for configuration files and metafiles for input.

- Man pages for the description of the `*.conf` file format removed.

### Changed

- Strings in JSON configuration files can now be longer than 256 characters.


## [5.6.13] - 2023-05-16

### Added


src/lpj/initdrain.c
/**************************************************************************************/
/**                                                                                \n**/
/**                 i  n  i  t  d  r  a  i  n  .  c                                \n**/
/**                                                                                \n**/
/**     C implementation of LPJmL                                                  \n**/
/**                                                                                \n**/
/**     Function initializes irrigation neighbours and river routings              \n**/
/**                                                                                \n**/
/**     Mathematical description of the river routing algorithm can be             \n**/
/**     found in:                                                                  \n**/
/**     Rost, S., Gerten, D., Bondeau, A., Lucht, W., Rohwer, J.,                  \n**/
/**     Schaphoff, S.: Agricultural green and blue water consumption and           \n**/
/**     its influence on the global water system. Water Resources                  \n**/
/**     Research (in press).                                                       \n**/
/**                                                                                \n**/
/**     River routing is implemented using the Pnet library.                       \n**/
/**     Pnet  is described in:                                                     \n**/
/**     W. von Bloh, 2008. Sequential and Parallel Implementation of               \n**/
/**     Networks. In P. beim Graben, C. Zhou, M. Thiel, and J. Kurths              \n**/
/**     (eds.), Lectures in Supercomputational Neuroscience, Dynamics in           \n**/
/**     Complex Brain Networks, Springer, 279-318.                                 \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

static int *getindex(const Input_netcdf input,const Cell grid[],
                     const Config *config)
{
  int *item=NULL;
#ifdef USE_MPI
  int *vec;
  int *counts;
  int *offsets;
#endif
  int i,cell,n;
  int *index;
  Bool iserr;
  iserr=FALSE;
  if(isroot(*config))
  {
    item=newvec(int,config->nall);
    if(item==NULL)
      iserr=TRUE;
  }
#ifdef USE_MPI
  counts=newvec(int,config->ntask);
  if(counts==NULL)
    iserr=TRUE;
  offsets=newvec(int,config->ntask);
  if(offsets==NULL)
    iserr=TRUE;
  vec=newvec(int,config->ngridcell);
  if(vec==NULL)
    iserr=TRUE;
  if(iserror(iserr,config))
  {
    if(isroot(*config))
      free(item);
    free(vec);
    free(counts);
    free(offsets);
    return NULL;
  }
  getcounts(counts,offsets,config->nall,1,config->ntask);
  for(cell=0;cell<config->ngridcell;cell++)
    vec[cell]=(int)getindexinput_netcdf(input,&grid[cell].coord);
  MPI_Gatherv(vec,config->ngridcell,MPI_INT,
              item,counts,offsets,MPI_INT,
              0,config->comm);
  free(vec);
  free(counts);
  free(offsets);
#else
  if(iserr)
    return NULL;
  for(cell=0;cell<config->ngridcell;cell++)
    item[cell]=(int)getindexinput_netcdf(input,&grid[cell].coord);
#endif
  n=getindexsize_netcdf(input);
  index=newvec(int,n);
  if(index==NULL)
    iserr=TRUE;
  else if(isroot(*config))
  {
    for(i=0;i<n;i++)
      index[i]=-1;
    for(cell=0;cell<config->nall;cell++)
    {
      if(item[cell]>=n)
      {
        fprintf(stderr,"ERROR202: Invalid index %d at cell %d in '%s'.\n",
                item[cell],cell+config->startgrid,
                config->neighb_irrig_filename.name);
        iserr=TRUE;
        break;
      }
      index[item[cell]]=cell;
    } 
    free(item);
  }
  if(iserror(iserr,config))
  {
    free(index);
    return NULL;
  }
#ifdef USE_MPI
  MPI_Bcast(index,n,MPI_INT,0,config->comm);
#endif
  return index;
} /* of 'getindex' */

static Bool getroute(FILE *file, /**< pointer to binary file */
                     Routing *r, /**< River route read */
                     Bool swap   /**< Byte order has to be changed */
                    )            /** \return TRUE on error */
{
  /* Function reads river route from file */
  if(fread(r,sizeof(Routing),1,file)!=1)
    return TRUE;
  if(swap)
  {
    /* byte order has to be changed */
    r->index=swapint(r->index);
    r->len=swapint(r->len);
  }
  return FALSE;
} /* of 'getroute' */

static Bool initirrig(Cell grid[],    /**< Cell grid             */
                      Config *config  /**< LPJ configuration     */
              )                       /** \return TRUE on error */
{
  Infile irrig_file;
  String line;
  int cell,neighb_irrig,rc,*index=NULL,n;
  /* open neighbour irrigation file */
  irrig_file.fmt=config->neighb_irrig_filename.fmt;
  if(openinputdata(&irrig_file,&config->neighb_irrig_filename,"irrigation",NULL,LPJ_INT,1.0,config))
    return TRUE;
  if(config->neighb_irrig_filename.fmt==CDF)
  {
    index=getindex(irrig_file.cdf,grid,config);
    if(index==NULL)
    {
      closeinput(&irrig_file);
      return TRUE;
    }
    n=getindexsize_netcdf(irrig_file.cdf);
  }
  /* initialize pnet structure for irrigation network */
#ifdef USE_MPI
  config->irrig_neighbour=pnet_init(config->comm,
                                    (sizeof(Real)==sizeof(double)) ? MPI_DOUBLE : MPI_FLOAT,
                                    config->nall);
#else
  config->irrig_neighbour=pnet_init(sizeof(Real),config->nall);
#endif
  if(config->irrig_neighbour==NULL)
  {
    fputs("ERROR140: Cannot initialize irrigation network.\n",stderr);
    fflush(stderr);
    free(index);
    closeinput(&irrig_file);
    return TRUE;
  }
  for(cell=0;cell<config->ngridcell;cell++)
  {
    /* read connection from file */
    if(readintinputdata(&irrig_file,&neighb_irrig,NULL,&grid[cell].coord,cell+config->startgrid,&config->neighb_irrig_filename))
    {
      free(index);
      closeinput(&irrig_file);
      return TRUE;
    }
    if(config->neighb_irrig_filename.fmt==CDF)
    {
      if(neighb_irrig<0 ||  neighb_irrig>=n)
      {
        fprintf(stderr,"ERROR203: Invalid irrigation neighbour %d of cell %d (%s).\n",
                neighb_irrig,cell+config->startgrid,sprintcoord(line,&grid[cell].coord));
        closeinput(&irrig_file);
        free(index);
        return TRUE;
      }
      neighb_irrig=index[neighb_irrig];
      if(neighb_irrig==-1)
      {
        fprintf(stderr,"ERROR203: Invalid irrigation neighbour %d of cell %d (%s).\n",
                neighb_irrig,cell+config->startgrid,sprintcoord(line,&grid[cell].coord));
        closeinput(&irrig_file);
        free(index);
        return TRUE;
      }
    }
    /* add connection to network */
    rc=pnet_addconnect(config->irrig_neighbour,
                       cell+config->startgrid-config->firstgrid,
                       neighb_irrig-config->firstgrid);
    if(rc)
    {
      fprintf(stderr,"ERROR142: Cannot add irrigation neighbour %d of cell %d: %s.\n",
              neighb_irrig,
              cell+config->startgrid,
              pnet_strerror(rc));
      fflush(stderr);
      closeinput(&irrig_file);
      free(index);
      return TRUE;
    }
  } /* of 'for(cell=...)' */
  closeinput(&irrig_file);
  free(index);
  config->irrig_back=pnet_dup(config->irrig_neighbour); /* copy network */
  return (config->irrig_back==NULL);
} /* of 'initirrig' */

static Bool initriver(Cell grid[],Config *config)
{
  Infile drainage,river;
  int cell,rc;
  Routing r;
  Header header;
  String headername,line;
  int *index,n,version,ncoeff;
  Real len;
  Bool missing;
  size_t offset;
  drainage.fmt=config->drainage_filename.fmt;
  if(config->drainage_filename.fmt==CDF)
  {
    drainage.cdf=openinput_netcdf(&config->drainage_filename,
                                  NULL,0,config);
    if(drainage.cdf==NULL)
      return TRUE;
    river.cdf=openinput_netcdf(&config->river_filename,"m",0,config);
    if(river.cdf==NULL)
    {
      closeinput_netcdf(drainage.cdf);
      return TRUE;
    }
    index=getindex(drainage.cdf,grid,config);
    if(index==NULL)
    {
      closeinput_netcdf(drainage.cdf);
      return TRUE;
    }
    n=getindexsize_netcdf(drainage.cdf);
  }
  else
  {

    if((drainage.file=openinputfile(&header,&drainage.swap,&config->drainage_filename,
                                    headername,&version,&offset,FALSE,config))==NULL)
      return TRUE;
    if(version>=3 && header.datatype!=LPJ_INT)
    {
      if(isroot(*config))
        fprintf(stderr,"ERROR217: Datatype %s in drainage file '%s' is not int.\n",typenames[header.datatype],config->drainage_filename.name);
      fclose(drainage.file);
      return TRUE;
    }
    if(config->drainage_filename.fmt!=RAW && header.nbands!=2)
    {
      if(isroot(*config))
        fprintf(stderr,"ERROR218: Number of bands=%d in drainage file '%s' is not 2.\n",
                header.nbands,config->drainage_filename.name);
      fclose(drainage.file);
      return TRUE;
    }
    if(isroot(*config) && config->drainage_filename.fmt!=META && getfilesizep(drainage.file)!=sizeof(Routing)*header.ncell+headersize(headername,version)+offset)
      fprintf(stderr,"WARNING032: File size of '%s' does not match nyear*ncell*nbands.\n",config->drainage_filename.name);
    /* seek startgrid positions in drainage file */
    if(fseek(drainage.file,sizeof(Routing)*(config->startgrid-header.firstcell)+offset,SEEK_CUR))
    {
      fprintf(stderr,"ERROR139: Cannot seek to drainage of cell %d.\n",
              config->startgrid);
      fclose(drainage.file);
      return TRUE;
    }
  }
  /* initialize pnet structure for drainage network */
#ifdef USE_MPI
  config->route=pnet_init(config->comm,
                          (sizeof(Real)==sizeof(double)) ? MPI_DOUBLE : MPI_FLOAT,
                          config->nall);
#else
  config->route=pnet_init(sizeof(Real),config->nall);
#endif

  if(config->route==NULL)
  {
    fputs("ERROR143: Cannot initialize river network.\n",stderr);
    fflush(stderr);
    closeinput(&drainage);
    return TRUE;
  }
  for(cell=0;cell<config->ngridcell;cell++)
  {
    if(config->drainage_filename.fmt==CDF)
    {
      if(readintinput_netcdf(drainage.cdf,&r.index,&grid[cell].coord,&missing) || missing)
      {
        closeinput_netcdf(drainage.cdf);
        closeinput_netcdf(river.cdf);
        free(index);
        return TRUE;
      }
      if(r.index<0 ||  r.index>=n)
      {
        fprintf(stderr,"ERROR203: Invalid irrigation neighbour %d of cell %d (%s).\n",
                r.index,cell+config->startgrid,sprintcoord(line,&grid[cell].coord));
        closeinput_netcdf(drainage.cdf);
        closeinput_netcdf(river.cdf);
        free(index);
        return TRUE;
      }
      r.index=index[r.index];
      if(r.index==-1)
      {
        fprintf(stderr,"ERROR203: Invalid irrigation neighbour %d of cell %d (%s).\n",
                r.index,cell+config->startgrid,sprintcoord(line,&grid[cell].coord));
        closeinput_netcdf(drainage.cdf);
        closeinput_netcdf(river.cdf);
        free(index);
        return TRUE;
      }
      if(readinput_netcdf(drainage.cdf,&len,&grid[cell].coord))
      {
        closeinput_netcdf(drainage.cdf);
        closeinput_netcdf(river.cdf);
        free(index);
        return TRUE;
      }
      /* calculate transfer function */
      grid[cell].discharge.tfunct=transfer_function(len,&ncoeff);
    }
    else
    {
      if(getroute(drainage.file,&r,drainage.swap))
      {
        fprintf(stderr,"ERROR144: Cannot read river route for cell %d.\n",
                cell+config->startgrid);
        fflush(stderr);
        fclose(drainage.file);
        return TRUE;
      }
      /* calculate transfer function */
      grid[cell].discharge.tfunct=transfer_function(r.len,&ncoeff);
    }
    grid[cell].discharge.next=r.index;

    if(grid[cell].discharge.tfunct==NULL)
    {
      printallocerr("tfunct");
      closeinput(&drainage);
      if(config->drainage_filename.fmt==CDF)
      {
        closeinput_netcdf(river.cdf);
        free(index);
      }
      return TRUE;
    }
    /* initialize delay queue with the same size */
    if(grid[cell].discharge.queue==NULL) /* has queue been read by freadcell? */
    {
      grid[cell].discharge.queue=newqueue(ncoeff); /* no, allocate it */
      if(grid[cell].discharge.queue==NULL)
      {
        printallocerr("queue");
        closeinput(&drainage);
        if(config->drainage_filename.fmt==CDF)
        {
          closeinput_netcdf(river.cdf);
          free(index);
        }
        return TRUE;
      }
    }
    else if(queuesize(grid[cell].discharge.queue)!=ncoeff) /* check whether queue size in restart file has the same length as ncoeff */
    {
      /* The queue size is different, so resize the queue and set it to zero */
      fprintf(stderr,"ERROR256: Size of discharge queue=%d of cell %d in restart file differs from %d, queue is resized and set to zero.\n",
             queuesize(grid[cell].discharge.queue),cell+config->startgrid,ncoeff);
      freequeue(grid[cell].discharge.queue);
      grid[cell].discharge.queue=newqueue(ncoeff);
      if(grid[cell].discharge.queue==NULL)
      {
        printallocerr("queue");
        closeinput(&drainage);
        if(config->drainage_filename.fmt==CDF)
        {
          closeinput_netcdf(river.cdf);
          free(index);
        }
        return TRUE;
      }
    }

    if(r.index>=0)
    {
      /* add connection */
      rc=pnet_addconnect(config->route,
                         cell+config->startgrid-config->firstgrid,
                         r.index-config->firstgrid);
      if(rc)
      {
        fprintf(stderr,"ERROR146: Cannot add river route from cell %d to cell %d in '%s': %s.\n",
                r.index,cell+config->startgrid,config->drainage_filename.name,pnet_strerror(rc));
        fflush(stderr);
        closeinput(&drainage);
        if(config->drainage_filename.fmt==CDF)
        {
          closeinput_netcdf(river.cdf);
          free(index);
        }
        return TRUE;
      }

/*    printf("c:%d next:%d size:%d d:%d first.id:%d\n",cell,
             grid[cell].discharge.next,grid[cell].discharge.queue.size,r.len,
             grid[cell].discharge.queue.first); */

    }

  } /* of for '(cell=0;..)' */

/*   printf("outsize:%d\n",config->route->outsize); */
/*   printf("lo:%d hi:%d\n",config->route->lo,config->route->hi); */

  closeinput(&drainage);
  if(config->drainage_filename.fmt==CDF)
  {
    closeinput_netcdf(river.cdf);
    free(index);
  }
  return FALSE;
} /* of 'initriver' */

Bool initdrain(Cell grid[],    /**< Cell grid             */
               Config *config  /**< LPJ configuration     */
              )                /** \return TRUE on error  */
{
  if(config->withlanduse!=NO_LANDUSE)
  {
    if(iserror(initirrig(grid,config),config))
      return TRUE; 
    pnet_reverse(config->irrig_neighbour);
    pnet_setup(config->irrig_neighbour);
    pnet_setup(config->irrig_back);
  }
  if(iserror(initriver(grid,config),config))
    return TRUE;
  pnet_reverse(config->route);
  pnet_setup(config->route);
  return FALSE;
} /* of 'initdrain' */

/*
case1: l==0, i.e. c==0 && d[1]==-9 no value
case2: l>0 && v>l, i.e. d[c+1]==-1 ocean or inland sink, whole water of considered cell drains to the ocean or inland sink
case3: l>0 && v<=l, i.e. d[c+1]>=0 || d[c+1]==-1 distribution of the water between two cells, or one cell and the ocean/inland sink

*/
